Fix several bugs in counter panel.

stellama0208 opened the pull request:
- Identify a counter selection by counter id, rather than the combination of name + start time ts.
- Query a counter selection by finding a row with fitted time range, rather than with the exact time.
- These change is to combat time quantization happening in counter track.
 - Bug 1: http://b/147922469, #5.
 - Bug 2: http://b/149959272. #1.
 - Bug 3: http://b/149959272, #2.

silence-do-good commented on gapic/src/main/com/google/gapid/perfetto/models/CounterTrack.java:
```diff
@@ -136,19 +145,21 @@ private String counterSQL() {
       res.forEachRow((i, r) -> {
         data.ts[i] = r.getLong(0);
         data.values[i] = r.getDouble(2);
+        data.ids[i] = r.getLong(3);
```
Can we remove the (1) element from the rangeSql if its not getting used anywhere?

stellama0208 replied:
It's actually used at `data.ts[rows] = res.getLong(rows - 1, 1, 0);`.

silence-do-good commented on gapic/src/main/com/google/gapid/perfetto/models/CounterTrack.java:
```diff
@@ -39,17 +40,21 @@
+      "where ts <= %d and ts + dur > %d";
   private static final String RANGE_SQL =
-      "select ts, ts + dur, value from %s " +
+      "select ts, ts + dur, value, id from %s " +
```
Is the ts + dur under select getting used? If not, we can remove that from the query.

stellama0208 replied:
It's actually used at `data.ts[rows] = res.getLong(rows - 1, 1, 0);`.

silence-do-good commented on gapic/src/main/com/google/gapid/perfetto/models/CounterTrack.java:
```diff
@@ -99,13 +104,15 @@ private String viewSql() {
       });
       data.ts[rows] = res.getLong(rows - 1, 1, 0);
       data.values[rows] = data.values[rows - 1];
+      data.ids[rows] = data.ids[rows - 1];
```
Is it possible to add a comment to why we are adding an extra index everywhere? Would help if any one of the functions had a comment.  Ignore this if its too trivial.

stellama0208 replied:
It's actually for transforming (ts, ts + dur, value) scheme to (ts, value) scheme. The added row is for the (very last timestamp - very last value) pair. 

Since this comes from the design of CounterTrack at the very beginning and done by Pascal, I might not be the best person to explain it. Will omit the comment for this PR, maybe we can re-check this thing when Pascal is back.

silence-do-good commented on gapic/src/main/com/google/gapid/perfetto/models/CounterTrack.java:
```diff
@@ -295,7 +313,7 @@ public Values combine(Values other) {
 
     @Override
-    public Selection<Values.Key> build() {
+    public Selection<Long> build() {
```
Delete the Key class below this function if its not getting used anywhere.

stellama0208 replied:
Thanks for reminding, forgot this bit! Done in the new change.

stellama0208 commented:
Addressed the comments above. PTAL!

Another change in the updated PR is that now I query counter selection based on id rather than time range, this will further solve another bug that you couldn't select counter when zoomed too **out** in a long trace.

silence-do-good commented on gapic/src/main/com/google/gapid/perfetto/models/CounterTrack.java:
```diff
@@ -158,44 +167,46 @@ private String rangeSql(TimeSpan ts) {
-            .boxed()
-            .forEach(t -> valueKeys.add(new Values.Key(name, t)));
+      for (long[] keys : ids) {
+        valueKeys.addAll(Arrays.stream(keys).boxed().collect(Collectors.toSet()));
```
Nit: Didn't notice this before. The old code seems to ignore the last dummy entry. I. don't think the new one does that. However, I'm not too sure of whether it will affect the end result or not.

silence-do-good commented on gapic/src/main/com/google/gapid/perfetto/models/CounterTrack.java:
```diff
@@ -158,44 +167,46 @@ private String rangeSql(TimeSpan ts) {
-            .boxed()
-            .forEach(t -> valueKeys.add(new Values.Key(name, t)));
+      for (long[] keys : ids) {
+        valueKeys.addAll(Arrays.stream(keys).boxed().collect(Collectors.toSet()));
```
Nit: Didn't notice this before. The old code seems to ignore the last dummy entry. I. don't think the new one does that. However, I'm not too sure of whether it will affect the end result or not.

rschiu replied:
It's probably better this way in the new code.  The old code use name + ts as the key. The last value is artificially created and doesn't actually exist in the database.  In the new code, the key is the id.  The last value is just a duplicate of the previous value.  Since it's being inserted into a Set, the duplicate just get ignored.

stellama0208 replied:
Yes, it's taking advantage of set+long to keep code shorter.

rschiu commented on gapic/src/main/com/google/gapid/perfetto/models/CounterTrack.java:
```diff
@@ -158,44 +167,46 @@ private String rangeSql(TimeSpan ts) {
-            .boxed()
-            .forEach(t -> valueKeys.add(new Values.Key(name, t)));
+      for (long[] keys : ids) {
+        valueKeys.addAll(Arrays.stream(keys).boxed().collect(Collectors.toSet()));
```
I think java.util.Collections.addAll would work and that would be best if it does.
If it doesn't, then you actually want to use forEach like before.  Using collect will unnecessary create a Set to contain the value of the stream.
Also, I don't think you need to boxed it if you are using forEach, but not 100% sure.

stellama0208 replied:
Rewrote it to avoid set creation.

rschiu commented on gapic/src/main/com/google/gapid/perfetto/models/CounterTrack.java:
```diff
@@ -39,17 +40,19 @@
 public class CounterTrack extends Track.WithQueryEngine<CounterTrack.Data> {
   private static final String VIEW_SQL_DELTA =
-      "select ts + 1 ts, lead(ts) over win - ts dur, lead(value) over win value " +
+      "select ts + 1 ts, lead(ts) over win - ts dur, lead(value) over win value, id " +
```
This should be 'lead(id) over win id' instead of id.

stellama0208 replied:
Nice catch, thanks! Done.
